tokio/src/sync/notify.rs

```diff
@@ -0,0 +1,496 @@
+/// [`Notify::recv`] waits for a permit to become available and
+/// [`Notify::notify_one`] sets a permit **if there currently are no available
+/// permits**.
+///
+/// If `notify_one` is called **before** `recv()`, then the next call to
+/// `recv()` will complete immediately, consuming the permit. Any subsequent
+/// calls to `recv()` will wait for a new permit.
+///
+/// If `notify_one` is called **multiple** times before `recv()`, only a
+/// **single** permit is stored. The next call to `recv()` will complete
+/// immediately, but the one after will wait for a new permit.
+///
+/// # Examples
+///
+/// Basic usage.
+///
+/// ```
+/// use tokio::sync::Notify;
+/// use std::sync::Arc;
+///
+/// #[tokio::main]
+/// async fn main() {
+///     let notify = Arc::new(Notify::new());
+///     let notify2 = notify.clone();
+///
+///     tokio::spawn(async move {
+///         notify2.recv().await;
+///         println!("received a notification");
+///     });
+///
+///     notify.notify_one();
+/// }
+/// ```
+///
+/// Unbound mpsc channel.
+///
+/// ```
+/// use tokio::sync::Notify;
+/// use std::collections::VecDeque;
+/// use std::sync::Mutex;
+///
+/// struct Channel<T> {
+///     values: Mutex<VecDeque<T>>,
+///     notify: Notify,
+/// }
+///
+/// impl<T> Channel<T> {
+///     pub fn send(&self, value: T) {
+///         self.values.lock().unwrap()
+///             .push_back(value);
+///
+///         // Notify the consumer a value is available
+///         self.notify.notify_one();
+///     }
+///
+///     pub async fn recv(&self) -> T {
+///         loop {
+///             // Drain values
+///             if let Some(value) = self.values.lock().unwrap().pop_front() {
+///                 return value;
+///             }
+///
+///             // Wait for values to be available
+///             self.notify.recv().await;
+///         }
+///     }
+/// }
+/// ```
+///
+/// [park]: std::thread::park
+/// [unpark]: std::thread::Thread::unpark
+/// [`Notify::recv`]: Notify::recv()
+/// [`Notify::notify_one`]: Notify::notify_one()
+/// [`Semaphore`]: crate::sync::Semaphore
+#[derive(Debug)]
+pub struct Notify {
+    state: AtomicU8,
+    waiters: Mutex<LinkedList<Waiter>>,
+}
+
+#[derive(Debug)]
+struct Waiter {
+    /// Waiting task's waker
+    waker: Option<Waker>,
+
+    /// `true` if the notification has been assigned to this waiter.
+    notified: bool,
+}
+
+#[derive(Debug)]
+struct RecvFuture<'a> {
+    /// The `Notify` being received on.
+    notify: &'a Notify,
+
+    /// The current state of the receiving process.
+    state: RecvState,
+
+    /// Entry in the waiter `LinkedList`.
+    waiter: linked_list::Entry<Waiter>,
+}
+
+#[derive(Debug)]
+enum RecvState {
+    Init,
+    Waiting,
+    Done,
+}
+
+/// Initial "idle" state
+const EMPTY: u8 = 0;
+
+/// One or more threads are currently waiting to be notified.
+const WAITING: u8 = 1;
+
+/// Pending notification
+const NOTIFIED: u8 = 2;
```

Comment:
Should these be
```rust
#[repr(u8)]
enum State {
    Empty = 0,
    Waiting = 1,
    Notified = 2,
}
```
So that we avoid the
```rust
_ => unreachable!()
```
match arms?

Reply:
In order to turn the `u8` _into_ a `State`, there basically still needs to be a fallible match against a u8 _somewhere_

Reply:
I tried, it seems of limited value and makes the code messier (more conversions).

https://github.com/tokio-rs/tokio/commit/7d38153bf235299286052010742eb6dd1511213f

I'm leaning towards no. Thoughts?